File: sdks/csharp/src/Verdocs.Sdk/Resources/ApiKeys.cs
using Verdocs.Models;

namespace Verdocs.Resources;

/// <summary>
/// API key calls, reached through the endpoint's ApiKeys property. API keys authenticate
/// server-to-server calls and act as one profile; every call requires the caller to be an
/// admin of the organization. Rotating or updating a key never invalidates sessions already
/// issued from it, so keys can be rotated at any time without disrupting an application.
/// </summary>
public sealed class ApiKeys
{
    private readonly VerdocsEndpoint _endpoint;

    internal ApiKeys(VerdocsEndpoint endpoint)
    {
        _endpoint = endpoint;
    }

    /// <summary>
    /// Gets the API keys for the caller's organization, each with its acting profile. Secrets
    /// are never included in listings; they are returned only by <see cref="CreateAsync"/>
    /// and <see cref="RotateAsync"/>.
    ///
    /// <example>
    /// <code>
    /// var apiKeys = await endpoint.ApiKeys.ListAsync();
    /// </code>
    /// </example>
    /// </summary>
    /// <param name="cancellationToken">Token to cancel the operation.</param>
    /// <returns>The organization's API keys, without secrets.</returns>
    /// <exception cref="VerdocsApiException">The call failed, for example because the caller is not an admin.</exception>
    /// <sdkOperation>apiKey.getApiKeys</sdkOperation>
    /// <sdkGroup>ApiKey</sdkGroup>
    /// <sdkPage>Endpoints</sdkPage>
    public async Task<IReadOnlyList<ApiKey>> ListAsync(CancellationToken cancellationToken = default)
    {
        return await _endpoint.SendAsync<List<ApiKey>>(HttpMethod.Get, "/v2/api-keys", null, cancellationToken)
            .ConfigureAwait(false);
    }

    /// <summary>
    /// Creates an API key acting as the given profile. Store the returned secret safely: this
    /// response and <see cref="RotateAsync"/> are the only places it appears.
    ///
    /// <example>
    /// <code>
    /// var apiKey = await endpoint.ApiKeys.CreateAsync(new CreateApiKeyRequest
    /// {
    ///     Name = "Integration key",
    ///     ProfileId = "d2338742-f3a1-465b-8592-806587413cc1",
    /// });
    /// </code>
    /// </example>
    /// </summary>
    /// <param name="request">Details for the new key.</param>
    /// <param name="cancellationToken">Token to cancel the operation.</param>
    /// <returns>The new key, including its secret.</returns>
    /// <exception cref="VerdocsApiException">The call failed, for example because the profile is not in the caller's organization.</exception>
    /// <sdkOperation>apiKey.createApiKey</sdkOperation>
    /// <sdkGroup>ApiKey</sdkGroup>
    /// <sdkPage>Endpoints</sdkPage>
    public Task<ApiKey> CreateAsync(CreateApiKeyRequest request, CancellationToken cancellationToken = default)
    {
        ArgumentNullException.ThrowIfNull(request);
        return _endpoint.SendAsync<ApiKey>(HttpMethod.Post, "/v2/api-keys", request, cancellationToken);
    }

    /// <summary>
    /// Rotates an API key's secret. Existing sessions issued from the key stay valid, so
    /// rotation is safe to do at any time.
    ///
    /// <example>
    /// <code>
    /// var rotated = await endpoint.ApiKeys.RotateAsync("d2338742-f3a1-465b-8592-806587413cc1");
    /// </code>
    /// </example>
    /// </summary>
    /// <param name="clientId">The client ID of the key to rotate.</param>
    /// <param name="cancellationToken">Token to cancel the operation.</param>
    /// <returns>The key with its new secret.</returns>
    /// <exception cref="VerdocsApiException">The call failed, for example because the key was not found.</exception>
    /// <sdkOperation>apiKey.rotateApiKey</sdkOperation>
    /// <sdkGroup>ApiKey</sdkGroup>
    /// <sdkPage>Endpoints</sdkPage>
    public Task<ApiKey> RotateAsync(string clientId, CancellationToken cancellationToken = default)
    {
        ArgumentException.ThrowIfNullOrEmpty(clientId);
        return _endpoint.SendAsync<ApiKey>(
            HttpMethod.Post, "/v2/api-keys/" + Uri.EscapeDataString(clientId) + "/rotate", null, cancellationToken);
    }

    /// <summary>
    /// Updates an API key's name, acting profile, or admin flag. Only the fields set on the
    /// request are sent; fields omitted are left unchanged.
    ///
    /// <example>
    /// <code>
    /// var updated = await endpoint.ApiKeys.UpdateAsync(
    ///     "d2338742-f3a1-465b-8592-806587413cc1",
    ///     new UpdateApiKeyRequest { Name = "Integration key (renamed)" });
    /// </code>
    /// </example>
    /// </summary>
    /// <param name="clientId">The client ID of the key to update.</param>
    /// <param name="request">The changes to apply.</param>
    /// <param name="cancellationToken">Token to cancel the operation.</param>
    /// <returns>The updated key, without its secret.</returns>
    /// <exception cref="VerdocsApiException">The call failed, for example because the key was not found.</exception>
    /// <sdkOperation>apiKey.updateApiKey</sdkOperation>
    /// <sdkGroup>ApiKey</sdkGroup>
    /// <sdkPage>Endpoints</sdkPage>
    public Task<ApiKey> UpdateAsync(string clientId, UpdateApiKeyRequest request, CancellationToken cancellationToken = default)
    {
        ArgumentException.ThrowIfNullOrEmpty(clientId);
        ArgumentNullException.ThrowIfNull(request);
        return _endpoint.SendAsync<ApiKey>(
            HttpMethod.Patch, "/v2/api-keys/" + Uri.EscapeDataString(clientId), request, cancellationToken);
    }

    /// <summary>
    /// Deletes an API key. Sessions already issued from the key remain valid until they
    /// expire, but no new sessions can be created with it.
    ///
    /// <example>
    /// <code>
    /// await endpoint.ApiKeys.DeleteAsync("d2338742-f3a1-465b-8592-806587413cc1");
    /// </code>
    /// </example>
    /// </summary>
    /// <param name="clientId">The client ID of the key to delete.</param>
    /// <param name="cancellationToken">Token to cancel the operation.</param>
    /// <returns>A task that completes when the key is deleted.</returns>
    /// <exception cref="VerdocsApiException">The call failed, for example because the caller is not an admin.</exception>
    /// <sdkOperation>apiKey.deleteApiKey</sdkOperation>
    /// <sdkGroup>ApiKey</sdkGroup>
    /// <sdkPage>Endpoints</sdkPage>
    public Task DeleteAsync(string clientId, CancellationToken cancellationToken = default)
    {
        ArgumentException.ThrowIfNullOrEmpty(clientId);
        return _endpoint.SendVoidAsync(
            HttpMethod.Delete, "/v2/api-keys/" + Uri.EscapeDataString(clientId), null, cancellationToken);
    }
}

File: sdks/python/src/verdocs/resources/api_keys.py
"""API key operations (js-sdk: Organizations/ApiKeys.ts).

API keys authenticate server-to-server calls; never use them from client
code. Authenticate with a key via endpoint.auth (client-credentials flow)
and re-authenticate as tokens expire. Keys may be rotated at any time
without disturbing existing sessions. Every call here requires an admin or
owner of the organization.
"""

from __future__ import annotations

from typing import TYPE_CHECKING, Any

from ..models.core import ApiKey
from ..models.organizations import ApiKeyCreateParams, ApiKeyUpdateParams

if TYPE_CHECKING:
    from .._endpoint import AsyncVerdocsEndpoint, VerdocsEndpoint

_API_KEYS_PATH = "/v2/api-keys"


def _write_body(params: ApiKeyCreateParams | ApiKeyUpdateParams) -> dict[str, Any]:
    return params.model_dump(mode="json", exclude_unset=True)


class ApiKeys:
    """API key calls for a sync endpoint."""

    def __init__(self, endpoint: VerdocsEndpoint) -> None:
        self._endpoint = endpoint

    def list(self) -> list[ApiKey]:
        """Get the API keys for the caller's organization via GET /v2/api-keys.

        Secrets are never included in the list; they are only shown by
        create() and rotate(). Mirrors js-sdk getApiKeys.

        Returns:
            The organization's API keys, each with its profile joined.

        Raises:
            AuthenticationError: The endpoint has no valid user session.
            VerdocsAPIError: The API returned another non-2xx status.
            VerdocsConnectionError: The request never reached the API.

        @sdkOperation apiKey.getApiKeys
        @sdkGroup ApiKey
        @sdkPage Endpoints
        """
        response = self._endpoint._request("GET", _API_KEYS_PATH)
        return [ApiKey.model_validate(entry) for entry in response.json()]

    def create(self, params: ApiKeyCreateParams) -> ApiKey:
        """Create an API key via POST /v2/api-keys. Mirrors js-sdk createApiKey.

        Example:
            key = endpoint.api_keys.create(ApiKeyCreateParams(name="CI", profile_id="profile-1"))
            client_secret = key.client_secret  # only shown here and on rotate

        Args:
            params: Name and acting profile for the new key, plus global_admin
                to give it full access to the organization.

        Returns:
            The new API key, including its client_secret. Store the secret;
            it is not returned by list().

        Raises:
            NotFoundError: The profile does not exist in the caller's organization.
            VerdocsAPIError: The API returned another non-2xx status.
            VerdocsConnectionError: The request never reached the API.

        @sdkOperation apiKey.createApiKey
        @sdkGroup ApiKey
        @sdkPage Endpoints
        """
        response = self._endpoint._request("POST", _API_KEYS_PATH, json=_write_body(params))
        return ApiKey.model_validate(response.json())

    def rotate(self, client_id: str) -> ApiKey:
        """Rotate an API key's secret via POST /v2/api-keys/{client_id}/rotate.

        Existing server-to-server sessions keep working, so rotation is safe
        to do at any time. Mirrors js-sdk rotateApiKey.

        Args:
            client_id: The client ID of the key to rotate.

        Returns:
            The key with its new client_secret.

        Raises:
            NotFoundError: No such key in the caller's organization.
            VerdocsAPIError: The API returned another non-2xx status.
            VerdocsConnectionError: The request never reached the API.

        @sdkOperation apiKey.rotateApiKey
        @sdkGroup ApiKey
        @sdkPage Endpoints
        """
        response = self._endpoint._request("POST", f"{_API_KEYS_PATH}/{client_id}/rotate")
        return ApiKey.model_validate(response.json())

    def update(self, client_id: str, params: ApiKeyUpdateParams) -> ApiKey:
        """Update an API key's name, acting profile, or global_admin flag via PATCH /v2/api-keys/{client_id}.

        Mirrors js-sdk updateApiKey.

        Args:
            client_id: The client ID of the key to update.
            params: The fields to change.

        Returns:
            The updated key. The secret is not included.

        Raises:
            NotFoundError: No such key in the caller's organization.
            VerdocsAPIError: The API returned another non-2xx status.
            VerdocsConnectionError: The request never reached the API.

        @sdkOperation apiKey.updateApiKey
        @sdkGroup ApiKey
        @sdkPage Endpoints
        """
        response = self._endpoint._request("PATCH", f"{_API_KEYS_PATH}/{client_id}", json=_write_body(params))
        return ApiKey.model_validate(response.json())

    def delete(self, client_id: str) -> None:
        """Delete an API key via DELETE /v2/api-keys/{client_id}. Mirrors js-sdk deleteApiKey.

        The API answers with a status marker that nothing consumes, so this
        returns None and relies on exceptions for failure. Deleting an
        unknown key is a silent no-op server-side.

        Args:
            client_id: The client ID of the key to delete.

        Raises:
            VerdocsAPIError: The API returned a non-2xx status.
            VerdocsConnectionError: The request never reached the API.

        @sdkOperation apiKey.deleteApiKey
        @sdkGroup ApiKey
        @sdkPage Endpoints
        """
        self._endpoint._request("DELETE", f"{_API_KEYS_PATH}/{client_id}")


class AsyncApiKeys:
    """API key calls for an async endpoint."""

    def __init__(self, endpoint: AsyncVerdocsEndpoint) -> None:
        self._endpoint = endpoint

    async def list(self) -> list[ApiKey]:
        """Get the API keys for the caller's organization via GET /v2/api-keys.

        Secrets are never included in the list; they are only shown by
        create() and rotate(). Mirrors js-sdk getApiKeys.

        Returns:
            The organization's API keys, each with its profile joined.

        Raises:
            AuthenticationError: The endpoint has no valid user session.
            VerdocsAPIError: The API returned another non-2xx status.
            VerdocsConnectionError: The request never reached the API.
        """
        response = await self._endpoint._request("GET", _API_KEYS_PATH)
        return [ApiKey.model_validate(entry) for entry in response.json()]

    async def create(self, params: ApiKeyCreateParams) -> ApiKey:
        """Create an API key via POST /v2/api-keys. Mirrors js-sdk createApiKey.

        Example:
            key = await endpoint.api_keys.create(ApiKeyCreateParams(name="CI", profile_id="profile-1"))
            client_secret = key.client_secret  # only shown here and on rotate

        Args:
            params: Name and acting profile for the new key, plus global_admin
                to give it full access to the organization.

        Returns:
            The new API key, including its client_secret. Store the secret;
            it is not returned by list().

        Raises:
            NotFoundError: The profile does not exist in the caller's organization.
            VerdocsAPIError: The API returned another non-2xx status.
            VerdocsConnectionError: The request never reached the API.
        """
        response = await self._endpoint._request("POST", _API_KEYS_PATH, json=_write_body(params))
        return ApiKey.model_validate(response.json())

    async def rotate(self, client_id: str) -> ApiKey:
        """Rotate an API key's secret via POST /v2/api-keys/{client_id}/rotate.

        Existing server-to-server sessions keep working, so rotation is safe
        to do at any time. Mirrors js-sdk rotateApiKey.

        Args:
            client_id: The client ID of the key to rotate.

        Returns:
            The key with its new client_secret.

        Raises:
            NotFoundError: No such key in the caller's organization.
            VerdocsAPIError: The API returned another non-2xx status.
            VerdocsConnectionError: The request never reached the API.
        """
        response = await self._endpoint._request("POST", f"{_API_KEYS_PATH}/{client_id}/rotate")
        return ApiKey.model_validate(response.json())

    async def update(self, client_id: str, params: ApiKeyUpdateParams) -> ApiKey:
        """Update an API key's name, acting profile, or global_admin flag via PATCH /v2/api-keys/{client_id}.

        Mirrors js-sdk updateApiKey.

        Args:
            client_id: The client ID of the key to update.
            params: The fields to change.

        Returns:
            The updated key. The secret is not included.

        Raises:
            NotFoundError: No such key in the caller's organization.
            VerdocsAPIError: The API returned another non-2xx status.
            VerdocsConnectionError: The request never reached the API.
        """
        response = await self._endpoint._request("PATCH", f"{_API_KEYS_PATH}/{client_id}", json=_write_body(params))
        return ApiKey.model_validate(response.json())

    async def delete(self, client_id: str) -> None:
        """Delete an API key via DELETE /v2/api-keys/{client_id}. Mirrors js-sdk deleteApiKey.

        The API answers with a status marker that nothing consumes, so this
        returns None and relies on exceptions for failure. Deleting an
        unknown key is a silent no-op server-side.

        Args:
            client_id: The client ID of the key to delete.

        Raises:
            VerdocsAPIError: The API returned a non-2xx status.
            VerdocsConnectionError: The request never reached the API.
        """
        await self._endpoint._request("DELETE", f"{_API_KEYS_PATH}/{client_id}")
